micrometer-core/src/test/java/io/micrometer/core/instrument/binder/jvm/JvmThreadDeadlockMetricsTest.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2024 VMware, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.core.instrument.binder.jvm;
+
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.simple.SimpleMeterRegistry;
+import org.junit.jupiter.api.Test;
+
+import java.lang.management.ThreadMXBean;
+import java.util.concurrent.TimeUnit;
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ * Tests for {@link JvmThreadDeadlockMetrics}.
+ *
+ * @author Ruth Kurniawati
+ */
+class JvmThreadDeadlockMetricsTest {
+
+    @Test
+    void deadlockedThreadMetrics() {
+        MeterRegistry registry = new SimpleMeterRegistry();
+        new JvmThreadDeadlockMetrics().bindTo(registry);
+        createTemporarilyDeadlockedThread();
+        assertThat(registry.get("jvm.threads.deadlocked").gauge().value()).isEqualTo(2);
+        assertThat(registry.get("jvm.threads.deadlocked.monitor").gauge().value()).isEqualTo(2);
+    }
+
+    @Test
+    void getDeadlockedThreadCountWhenFindDeadlockedThreadsIsNullShouldWork() {
+        ThreadMXBean threadBean = mock(ThreadMXBean.class);
+        when(threadBean.findDeadlockedThreads()).thenReturn(null);
+        assertThat(JvmThreadDeadlockMetrics.getDeadlockedThreadCount(threadBean)).isEqualTo(0);
+    }
+
+    @Test
+    void getDeadlockedThreadCountWhenFindMonitorDeadlockedThreadsIsNullShouldWork() {
+        ThreadMXBean threadBean = mock(ThreadMXBean.class);
+        when(threadBean.findMonitorDeadlockedThreads()).thenReturn(null);
+        assertThat(JvmThreadDeadlockMetrics.getDeadlockedMonitorThreadCount(threadBean)).isEqualTo(0);
+    }
+
+    private void sleep(int seconds) {
+        try {
+            TimeUnit.SECONDS.sleep(seconds);
+        }
+        catch (InterruptedException ignored) {
+        }
+    }
+
+    private void createTemporarilyDeadlockedThread() {
+        // create two threads that are deadlocked for about 5 seconds
```

Comment:
Could we control this with something like CountDownLatch so we can have the same effect without the test always taking multiple seconds?

Reply:
I changed the test to use `CountDownLatch`. Unfortunately I have to use `ReentrantLock` instead of the monitor lock, so the the `jvm.threads.deadlocked` is non-zero but `jvm.threads.deadlocked.monitor` is zero